client/my-sites/controller.js

```diff
@@ -235,44 +246,78 @@ module.exports = {
+					sites.once( 'change', () => sites.fetched ? resolve() : reject() );
+				} else {
+					resolve();
+				}
+			} );
+		};
 
-			// if there was a redirect, we should terminate processing of next routes
-			// and let the redirect proceed
-			if ( ! selectionComplete ) {
-				return;
+		const hardFetchSites = () => new Promise( ( resolve, reject ) => {
+			sites.initialized = false;
+			sites.once( 'change', () => {
+				if ( sites.fetched ) {
+					sites.initialized = true;
+					return resolve();
+				}
+
+				reject();
+			} );
+
+			sites.fetch();
+		} );
+
+		let waitingNotice;
+		const validateAllVisibileSitesAreFetched = () => new Promise( resolve => {
+			if ( currentUser.visible_site_count !== sites.getVisible().length ) {
+				if ( ! waitingNotice ) {
+					waitingNotice = notices.info( i18n.translate( 'Finishing set up…' ), { showDismiss: false } );
+				}
+
+				return hardFetchSites().then(
+					validateAllVisibileSitesAreFetched,
+					validateAllVisibileSitesAreFetched
+				);
 			}
-		} else {
-			// if sites has fresh data and siteID is invalid
-			// redirect to allSitesPath
-			if ( sites.fetched || ! sites.fetching ) {
-				return page.redirect( allSitesPath );
+
+			if ( waitingNotice ) {
+				notices.removeNotice( waitingNotice );
 			}
 
-			let waitingNotice;
-			const selectOnSitesChange = () => {
-				// if sites have loaded, but siteID is invalid, redirect to allSitesPath
-				if ( sites.select( siteID ) ) {
-					sites.initialized = true;
-					onSelectedSiteAvailable( context );
-					if ( waitingNotice ) {
-						notices.removeNotice( waitingNotice );
-					}
-				} else if ( ( currentUser.visible_site_count !== sites.getVisible().length ) ) {
-					sites.initialized = false;
-					waitingNotice = notices.info( i18n.translate( 'Finishing set up…' ), { showDismiss: false } );
-					sites.once( 'change', selectOnSitesChange );
-					sites.fetch();
-				} else {
-					page.redirect( allSitesPath );
+			resolve();
+		} );
+
+		// ensure we have fetched sites
+		const sitesFetchedPromise = waitForInProgressSitesFetch().then( validateAllVisibileSitesAreFetched );
+
+		// Select a site
+		sitesFetchedPromise
+			.catch( error => {
+				page.redirect( allSitesPath );
+				return Promise.reject( error );
+			} )
+			.then( () => {
+				if ( ! sites.select( siteID ) ) {
+					return page.redirect( allSitesPath );
 				}
-			};
-			// Otherwise, check when sites has loaded
-			sites.once( 'change', selectOnSitesChange );
-		}
+
+				const selectionContext = {
+					reactContext: context,
+					selectedSite: sites.getSelectedSite()
+				};
+
+				feedReduxStoreWithSelectedSite( selectionContext );
+
+				if ( isSelectedSiteDomainOnlySite( context.store.getState() ) && ! isPathAllowedForDomainOnlySite( selectionContext ) ) {
+					return renderSelectedSiteIsDomainOnly( selectionContext )
+				}
+
+				setRecentSitesPreferenceInReduxStore( selectionContext );
+			} );
+
 		next();
```

Comment:
@yurynix To better understand this problem, please let me know if this is correct (from a high level), and if not, what I'm missing here.

- If a site appears in the sitepicker, it has been fetched at some point, and we already know whether or not it is a domain-only site. There is a chance that the data could be stale, but since there isn't a way to convert a domain-only site to a normal site (and even if there was it wouldn't happen frequently), we can be confident that the `is_domain_only` option is accurate even for a stale `sites-list`.
- Since we already know if a domain is a domain-only site, we should be able to display the `This feature is not available for domains` when switching to any domain-only site from any sitepicker.
- Accessing a route like `/stats/day/:site` directly when the user does not have `sites-list` cached will display the loading state for that route (the stats page loading state, in this example), which is erroneous for domain-only sites. However, the only way that a user would see this is if they cleared their local storage and visited the route directly, so it is enough of an edge case to not handle in this PR.